### PR TITLE
WIP Transform RVs, not just their logp.

### PR DESCRIPTION
**Pull request by twiecki**
```python
@pm.model
def t_test():
    sigma = pm.HalfNormal('sigma', 1)
    mu = pm.Normal('mu', 0, sigma)

model = t_test.configure()
func = model.make_log_prob_function()
sd = -tf.ones((1,))
mu = tf.ones((1,))
func(sd, mu)
```

```
<tf.Tensor: id=325, shape=(), dtype=float32, numpy=-4.5149107>
```

### CODE REVIEW

File: pymc4/random_variables/random_variable.py [hunk 1]
```diff
@@ -157,6 +159,15 @@ def log_prob(self):
         """
         return self._transformed_distribution.log_prob(self)
 
+    def as_tensor(self):
```
Comment (ericmjl) [inline pymc4/random_variables/random_variable.py]: Can we leave a brief docstring here explaining why `as_tensor` is reimplemented here?

**Comment by twiecki** [comment]
Added a new commit with a new idea: We *always* transform distributions, just in some cases that transformation is the identity. This will save us a whole bunch of extra checks and unify the whole thing.

File: pymc4/random_variables/random_variable.py [hunk 2]
```diff
@@ -108,30 +108,39 @@ class RandomVariable(WithBackendArithmetic):
...
-        self._distribution = self._base_dist(name=name, *args, **kwargs)
+        self._untransformed_distribution = self._base_dist(name=name,
+                                                           *args, **kwargs)
         self._sample_shape = ()
         self._dim_names = ()
         self.name = name
         ctx = contexts.get_context()
         self._creation_context_id = id(ctx)
         self._backend_tensor = None
+        if "bijector" in kwargs.keys():
```
Comment (canyon289) [inline pymc4/random_variables/random_variable.py]: Being picky here but you could change this to self._bijector = kwargs.get("bijector",  _bijector)

Comment (twiecki) [reply pymc4/random_variables/random_variable.py]: we could do `self._bijector = kwargs.get("bijector", self._bijector)` if that's what you meant.

Comment (canyon289) [reply pymc4/random_variables/random_variable.py]: Yea that :)

File: pymc4/random_variables/random_variable.py [hunk 3]
```diff
@@ -140,62 +148,21 @@ def as_tensor(self):
         if self._backend_tensor is None:
             self._backend_tensor = ctx.var_as_backend_tensor(self)
 
-        return self._backend_tensor
+        return self._bijector.forward(self._backend_tensor)
 
 
 class ContinuousRV(RandomVariable):
```
Comment (canyon289) [inline pymc4/random_variables/random_variable.py]: If this and `DiscreteRV` both do nothing let's get rid of them. The extra indirection doesn't help any

Comment (twiecki) [reply pymc4/random_variables/random_variable.py]: Good point.

File: pymc4/random_variables/random_variable.py (hunk 1 above)
Comment (ericmjl) [reply pymc4/random_variables/random_variable.py]: nvm, I'm being dumb here.

**Comment by ericmjl** [comment]
@twiecki I pushed up a few more changes. Please check them out.

**Comment by twiecki** [comment]
@ericmjl Thanks! I accidentally removed them during a rebase but reimplemented them.

**Comment by twiecki** [comment]
Superseded by https://github.com/pymc-devs/pymc4/pull/92.

**Pull request closed without merging**